### ISSUE TITLE
About the number of variants selected into final model

### ISSUE THREAD

**Issue by felixhu93** [privefl/bigsnpr#186]
I have a simple question about the final number of variants these two methods choose in the final model:
1. For best C+T prediction model: I find the following codes:
##################
ind.keep <- unlist(purrr::map(all_keep, max_prs$id))
sum(lpval[ind.keep] > max_prs$thr.lp)
##################
so is the output of the second line the final number of variants (after thresholding) we use in our model?

2. For the LDpred2 method: Similarly, are the nonzero variants in "best_grid_sp" the final set of variants we use in our model?

Thank you so much!

**Comment by privefl** [comment]
1. Yes it is. You should be able to quickly verify this by looking directly at the resulting vector of effects.

2. Yes

**Comment by felixhu93** [comment]
Thank you so much! So if I calculate the R2 in the final linear model phenotype~PRS, are they the number of variables we can used to calculate adjusted-R2?

**Comment by privefl** [comment]
I think that in the adjusted-R2, the PRS is considered as a unique variable, not many.

**Comment by felixhu93** [comment]
Hi, I have follow up questions about the C+T methods code:

1. For the following code:

```
grid2$auc <- big_apply(multi_PRS, a.FUN = function(X, ind, s, y.train) {
  single_PRS <- rowSums(X[, ind + s * (0:2)])  ## replace by 0:21 in real data
  bigstatsr::AUC(single_PRS, y.train)
}, ind = 1:s, s = s, y.train = y[ind.train],
a.combine = 'c', block.size = 1, ncores = NCORES)
```

I know in this case, you use a 0-1 binary trait so we calculate AUC as our metric to select parameters. If I have **a continous trait** like height, can I change it to like this:
```
grid$r2 <- big_apply(multi_PRS, a.FUN = function(X, ind, s, y.train) {
  single_PRS <- rowSums(X[, ind + s * (0:2)])
  dt = data.frame(y=y.train, single_PRS, pc[ind.train,])
  lm.fit = lm(y~., data = dt)
  summary(lm.fit)$r.squared
}, ind = 1:s, s = s, y.train = y[ind.train],
a.combine = 'c', block.size = 1, ncores = NCORES)
```
(where pc is the Principle Comp matrix), and select parameters by R2?

2. The reason why I ask the number of variants questions is that when I try the C+T method on a height summary statistics and a target data with height as phenotype, the R2 is very low for testing data (0.04) while ind.keep <- ```
unlist(purrr::map(all_keep, max_prs$id))
sum(lpval[ind.keep] > max_prs$thr.lp)
```
return a number of variant more than 100,000. I know this question might not be approriate to be asked here, but combined with question 1, do you think when we are dealing with continous triat like height, R2 is a good metrics to select threshold and parameter settings?

Thank you so much!

**Comment by privefl** [comment]
Yes, you need to use an alternative metric for continuous traits.

However, I wouldn't use r2 as it is not signed. I would use the t-score corresponding to the PRS in `lm()`, or bigstatsr::pcor() (which should be a monotonic transformation of this t-score I guess).

What do you mean r2 is very low for height? In the UKBB I get 40-42% for pcor()^2.

**Comment by felixhu93** [comment]
Really appreciate to your quick response!
I mean, after I get the optimal parameters settings, calculate the PRS by using the code provided in tutorial
```
grid$r2 <- big_apply(multi_PRS, a.FUN = function(X, ind, s, y.train) {
    single_PRS <- X[, ind]
    dt = data.frame(y=y.train, x=single_PRS)
    lm.fit = lm(y~x, data = dt)
    coef(summary(lm.fit))[, "t value"][2]
  }, ind = 1:s, s = s, y.train = y[ind.train],
  a.combine = 'c', block.size = 1, ncores = NCORES)
  max_prs <- grid[which.max(grid$r2), ]
  ind.keep <- unlist(purrr::map(all_keep, max_prs$id))
  ind.ct <- ind.keep[lpval[ind.keep] > max_prs$thr.lp]
  opt.prs <- snp_PRS(G, beta[ind.keep], ind.keep = ind.keep, ind.test = ind.test,
                     lpS.keep = lpval[ind.keep], thr.list = max_prs$thr.lp)
```
and fitting the final linear model like this 
```
  dt <- data.frame(y[ind.test], opt.prs)
  lm.fit <- lm(y[ind.test]~., data = dt)
  R.ct.prs <- summary(lm.fit)$r.squared
```
the final r2 is pretty low, nearly to 4%, which I think fitting between the prs is not very good. Is this mean we also cannot use r2 as the metrics between final fitting?

**Comment by privefl** [comment]
Any update on this?
